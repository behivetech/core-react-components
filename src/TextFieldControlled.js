// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {omit} from 'lodash';

// Components
import TextField from './TextField.js';
import withFormControl from './withFormControl';
// TODO: Set up withValidation file.
const withValidation = (Component) => Component;

class TextFieldControlled extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: props.value || props.formState.getFieldValue(props.name) || '',
        };
        this.handleChange = this.handleChange.bind(this);
    }

    componentDidMount() {
        const {formState, name, value} = this.props;
        
        if (value) {
            formState.setFieldValue(name, value);
        }
    }

    handleChange(event) {
        const value = event.target.value;
        const {name, onChange, formState} = this.props;

        this.setState({value: value});
        formState.setFieldValueDebounced(name, value);
        onChange(event);
    }

    render() {
        const textField = (
            <TextField 
                {...omit(this.props, ['formState'])} 
                onChange={this.handleChange} 
                value={this.state.value} 
            />
        );

        return (this.props.validate)
            ? withValidation(textField)
            : textField;
    }
}

TextFieldControlled.propTypes = {
    formState: PropTypes.shape({
        /** Gets the field value from the FormControlled component */
        getFieldValue: PropTypes.func.isRequired,
        /** Sets the field value in the FormControlled to be a part of the JSON that will be pushed. */
        setFieldValue: PropTypes.func.isRequired,
        /** Same as setFieldValue, but deboinces the onChange events. */
        setFieldValueDebounced: PropTypes.func.isRequired,
    }),
    /** Name of the field. Required to be used as a key as part of the JSON that will be pushed */
    name: PropTypes.string.isRequired, 
    /** Callback function for additional functionality when the component has a change */
    onChange: PropTypes.func,
    /** Value of the component when it is checked. Defaults to true/false if nothing is set */
    value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.string,
        PropTypes.number,
    ]),
};

TextFieldControlled.defaultProps = {
    onChange: () => null,
    value: '',
};


export default withFormControl(TextFieldControlled);