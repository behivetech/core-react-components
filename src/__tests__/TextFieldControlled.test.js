import TextFieldControlled from '../TextFieldControlled';

jest.mock('../withFormControl');
jest.mock('../withValidation');

describe('TextFieldControlled', () => {
    const getEvent = (value='mock value') => {
        return {
            target: {value},
        };
    };
    const getProps = () => {
        return {
            className: 'mock-class',
            label: 'mock-label',
            name: 'mockName',
        };
    };

    it('should render with the correct props', () => {
        const wrapper = shallow(<TextFieldControlled {...getProps()} value="mock value" />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.dive()).toMatchSnapshot();
        expect(wrapper.props().onChange(getEvent())).toBe(null);
        wrapper.setProps({validate: {mockValidate: jest.fn()}});
        expect(wrapper.dive()).toMatchSnapshot();
        expect(wrapper.dive().instance().state.value).toBe('mock value');
    });

    it('should set state, formState and run onChange prop when changed', () => {
        const mockOnChange = jest.fn();
        const mockEvent = getEvent();
        const wrapperInstance = shallow(
            <TextFieldControlled {...getProps()} onChange={mockOnChange} />
        ).dive().instance();
        const {formState, name} = wrapperInstance.props;

        wrapperInstance.handleChange(mockEvent);
        expect(wrapperInstance.state.value).toBe('mock value');
        expect(formState.setFieldValueDebounced).toBeCalledWith(name, mockEvent.target.value);
        expect(mockOnChange).toBeCalledWith(mockEvent);
    });

    it('should set form state value if a value prop is set', () => {
        const wrapperInstance = shallow(
            <TextFieldControlled {...getProps()} value="mock value" />
        ).dive().instance();
        const {formState, name, value} = wrapperInstance.props;

        expect(formState.setFieldValue).toBeCalledWith(name, value);
    });
});