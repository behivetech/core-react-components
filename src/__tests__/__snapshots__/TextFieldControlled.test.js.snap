// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFieldControlled should render with the correct props 1`] = `
<TextFieldControlled
  className="mock-class"
  formState={
    Object {
      "formDisabled": false,
      "getFieldError": [MockFunction],
      "getFieldValue": [MockFunction],
      "initialSubmit": false,
      "setFieldValidators": [MockFunction],
      "setFieldValue": [MockFunction],
      "setFieldValueDebounced": [MockFunction],
      "shouldDisable": [MockFunction],
    }
  }
  label="mock-label"
  name="mockName"
  onChange={[Function]}
  value="mock value"
/>
`;

exports[`TextFieldControlled should render with the correct props 2`] = `
<TextField
  className="mock-class"
  label="mock-label"
  name="mockName"
  onChange={[Function]}
  value="mock value"
/>
`;

exports[`TextFieldControlled should render with the correct props 3`] = `
<FormFieldValidation
  className="mock-class"
  component={[Function]}
  formState={
    Object {
      "formDisabled": false,
      "getFieldError": [MockFunction],
      "getFieldValue": [MockFunction],
      "initialSubmit": false,
      "setFieldValidators": [MockFunction],
      "setFieldValue": [MockFunction] {
        "calls": Array [
          Array [
            "mockName",
            "mock value",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
      "setFieldValueDebounced": [MockFunction],
      "shouldDisable": [MockFunction],
    }
  }
  label="mock-label"
  name="mockName"
  onChange={[Function]}
  validate={
    Object {
      "mockValidate": [MockFunction],
    }
  }
  value="mock value"
/>
`;
